sdks/java/io/tika/src/main/java/org/apache/beam/sdk/io/tika/TikaSource.java

```diff
@@ -0,0 +1,466 @@
+      checkState(currentReader != null, "Call start() before advance()");
+      if (currentReader.advance()) {
+        return true;
+      }
+      return startNextNonemptyReader();
+    }
+
+    private boolean startNextNonemptyReader() throws IOException {
+      while (fileReadersIterator.hasNext()) {
+        currentReader = fileReadersIterator.next();
+        if (currentReader.start()) {
+          return true;
+        }
+        currentReader.close();
+      }
+      return false;
+    }
+
+    @Override
+    public String getCurrent() throws NoSuchElementException {
+      return currentReader.getCurrent();
+    }
+
+    @Override
+    public Instant getCurrentTimestamp() throws NoSuchElementException {
+      return currentReader.getCurrentTimestamp();
+    }
+
+    @Override
+    public void close() throws IOException {
+      if (currentReader != null) {
+        currentReader.close();
+      }
+      while (fileReadersIterator.hasNext()) {
+        fileReadersIterator.next().close();
+      }
+    }
+
+    @Override
+    public TikaSource getCurrentSource() {
+      return source;
+    }
+  }
+
+  static class TikaReader extends BoundedReader<String> {
+    private ExecutorService execService;
+    private final ContentHandlerImpl tikaHandler = new ContentHandlerImpl();
+    private String current;
+    private TikaSource source;
+    private String filePath;
+    private TikaIO.Read spec;
+    private org.apache.tika.metadata.Metadata tikaMetadata;
+    private Iterator<String> metadataIterator;
+
+    TikaReader(TikaSource source, String filePath) {
+      this.source = source;
+      this.filePath = filePath;
+      this.spec = source.getTikaInputRead();
+    }
+
+    @Override
+    public boolean start() throws IOException {
+      final InputStream is = TikaInputStream.get(Paths.get(filePath));
+      TikaConfig tikaConfig = null;
+      if (spec.getTikaConfigPath() != null) {
+        try {
+          tikaConfig = new TikaConfig(spec.getTikaConfigPath().get());
+        } catch (TikaException | SAXException e) {
+          throw new IOException(e);
+        }
+      }
+      final Parser parser = tikaConfig == null ? new AutoDetectParser()
+          : new AutoDetectParser(tikaConfig);
+      final ParseContext context = new ParseContext();
+      context.set(Parser.class, parser);
+      tikaMetadata = spec.getInputMetadata() != null ? spec.getInputMetadata()
+          : new org.apache.tika.metadata.Metadata();
+
+      if (spec.getMinimumTextLength() != null) {
+        tikaHandler.setMinTextLength(spec.getMinimumTextLength());
+      }
+
+      if (!Boolean.TRUE.equals(spec.getParseSynchronously())) {
+        // Try to parse the file on the executor thread to make the best effort
+        // at letting the pipeline thread advancing over the file content
+        // without immediately parsing all of it
+        execService = Executors.newFixedThreadPool(1);
+        execService.submit(new Runnable() {
```

Comment:
Oh. I thought this was a Tika feature. Why do we need this feature? If this is for returning results via a parsing callback, then we could have done this using a ParDo which would, upon getting the callback, do c.output(): this is a poor fit for the Source API.

Reply:
The way it works right now: Beam thread enters TikaReader, gets a piece of content, pushes it to the pipeline, it's not tied to the parser operations; with ParDo - the thread enters it and it will not leave ParDo untill all the Nmb file has been parsed, this beam thread will parse all the file itself; unless I'm missing something, IMHO the current implementation is more optimal.
There are some formats where the content will be reported in a single event, only in such cases ParDo will be OK, but we don't know at the generic level when it can be the case, we just let the users to do the sync parsing optionally...